nx/guides/advanced/complex_fft.livemd

```diff
@@ -0,0 +1,297 @@
+
+```elixir
+defmodule Signal do
+  import Nx.Defn
+  import Nx.Constants, only: [pi: 0]
+
+  defn source(t) do
+    f1 = 5; f2 = 20;
+    Nx.sin(2 * pi() * f1 * t ) + 1/2 * Nx.sin(2 * pi() * f2 * t)
+  end
+
+  defn sample(opts) do
+    fs = opts[:fs]
+    duration = opts[:duration]
+    sampling = Nx.linspace(0, duration, n: fs, endpoint: false, type: {:f, 32})
+    source(sampling)
+  end
+end
+```
+
+We sample our signal at fs=50Hz during 1s:
+
+```elixir
+fs = 50; duration= 1
+
+sample = Signal.sample(fs: fs, duration: 1)
+```
+
+### Analyse the signal with DFT
+
+The DFT algorithm will **approximates the original signal**. It gives us the following information:
+
+> The number at the index $i$ of the DFT results gives an approximation of the amplitude and phase of the sampled signal at the frequency $i$.
+
+In other words, doing `Nx.fft(sample)` returns a list of numbers indexed by the frequency.
+
+These are complex numbers as they carry the amplitude and phase information.
+
+```elixir
+dft = Nx.fft(sample)
+```
+
+Our frequency resolution - the accuracy of the DFT algorithm - is limited by the Nyquist rate, which is half the sampling rate, thus 25 Hz.
+
+Check [Nyquist](https://en.wikipedia.org/wiki/Nyquist_rate) to read more.
+
+We will therefor limit our study points to the indices below this threshold value by slicing the tensor, thus retain the 25 first ones.
+
+We are interested in the amplitudes only here. This is calculated by taking the norm of each number (`Nx.abs`).
+
+```elixir
+n = Nx.size(dft)
+
+# Nyquist threshold:
+max_freq_index = div(n, 2)
+
+# thus we keep only the meaningfull values and calculate the amplitudes
+amplitudes =  Nx.abs(dft)[0..max_freq_index]
+
+# the frequency bins, "n" of them spaced by fs/n=1 unit:
+frequencies = NxSignal.fft_frequencies(fs, fft_length: n)[0..max_freq_index]
+
+data1 = %{
+  frequencies: Nx.to_list(frequencies),
+  amplitudes: Nx.to_list(amplitudes)
+}
+
+VegaLite.new(width: 700, height: 300)
+|> VegaLite.data_from_values(data1)
+|> VegaLite.mark(:bar)
+|> VegaLite.encode_field(:x, "frequencies",
+  type: :quantitative,
+  title: "frequency (Hz)",
+  scale: [domain: [0, 50]]
+)
+|> VegaLite.encode_field(:y, "amplitudes",
+  type: :quantitative,
+  title: "amplitutde",
+  scale: [domain: [0, 30]]
+)
+```
+
+We see the peaks at 5Hz, 20Hz with their amplitudes (the second is half the first).
+
+This is indeed our incomming signal 🎉
+
+We can confirm this visual inspection with a peek into our data. We use `Nx.top_k` function.
+
+```elixir
+{values, indices} = Nx.top_k(amplitudes, k: 5)
+
+filtered_top_frequencies = fn values, indices, threshold ->
+  v = Nx.to_list(values)
+  v0 = Enum.at(v,0)
+  Enum.zip(v, Nx.to_list(indices))
+  |> Enum.filter(fn {v,_idx} -> v >=  v0 * threshold end)
+  |> Enum.unzip()
+  |> elem(1)
+
+end
+
+IO.puts("The top frequencies are: #{inspect(filtered_top_frequencies.(values, indices, 0.001 ))}")
```

Comment:
```suggestion
```

Reply:
Useless? maybe...

Reply:
Ah, I forgot to explain. I think just discussing over the output is better than inducing readers into falling back to Elixir at the first sign of trouble, if it makes sense. Sure, you can't determine things automatically, but even your solution uses prior knowledge anyway

Reply:
hehehe, trouble, kind of! Just that was tempted to use `top_k` but needs filtering.